src/main/java/se/liu/ida/hefquin/engine/data/mappings/impl/SchemaMappingImpl.java

```diff
@@ -1,44 +1,286 @@
+				graph.addSubPattern( resultTriple );
+			}
+		}
+
+		return graph;
 	}
 
 	@Override
+	/**
+	 * Translate solution mappings that using local vocabulary to global vocabularies
+	 */
 	public Set<SolutionMapping> applyToSolutionMapping( final SolutionMapping sm ) {
-		// TODO: implement this function
-		return Collections.singleton(sm);
+		return applyMapToSolutionMapping(sm, false);
 	}
 
 	@Override
 	public Set<SolutionMapping> applyInverseToSolutionMapping( final SolutionMapping sm ) {
-		// TODO: implement this function
-		return Collections.singleton(sm);
+		return applyMapToSolutionMapping(sm, true);
+	}
+
+	protected Set<SolutionMapping> applyMapToSolutionMapping(final SolutionMapping sm, final boolean Inverse ) {
+		final Binding binding = sm.asJenaBinding();
+		// If the given solution mapping is the empty mapping, simply return it unchanged.
+		if ( binding.isEmpty() ) {
+			return Collections.singleton(sm);
+		}
+
+		Set<SolutionMapping> resultsSolMaps = new HashSet<>();
+		Set<SolutionMapping> tmpSolMaps = new HashSet<>();
+
+		final SolutionMapping initialSol = new SolutionMappingImpl( BindingBuilder.create().build() );
+		resultsSolMaps.add(initialSol);
+		final Iterator<Var> it = binding.vars();
+		while ( it.hasNext() ) {
+			final Var v = it.next();
+			Set<Node> terms;
+			if ( Inverse ) {
+				terms = applyMapToGlobalURI(binding.get(v), g2lMap);
+			}
+			else {
+				terms = applyMapToLocalURI(binding.get(v), l2gMap);
+			}
+
+			for ( Node uri: terms ) {
+				final BindingBuilder newBinding = BindingBuilder.create();
+				newBinding.add(v, uri);
+				final SolutionMapping sol = new SolutionMappingImpl( newBinding.build() );
+				for ( SolutionMapping tmpSol : resultsSolMaps ){
+					tmpSolMaps.add( SolutionMappingUtils.merge(sol, tmpSol) );
+				}
+			}
+			resultsSolMaps.clear();
+			resultsSolMaps.addAll( tmpSolMaps );
+			tmpSolMaps.clear();
+		}
+		return resultsSolMaps;
+	}
+
+	public Set<Node> applyMapToLocalURI( final Node node, final Map<Node, Set<TermMapping>> vocMap ) {
+		final Set<TermMapping> termMappings = vocMap.get(node);
+		final Set<Node> results = new HashSet<>();
+		if ( termMappings != null ) {
+			for (TermMapping mappingTypes : termMappings) {
+				results.addAll( mappingTypes.getTranslatedTerms() );
+			}
+		}
+		else {
+			results.add(node);
+		}
+
+		return results;
+	}
+
+	/**
+	 * Only apply equivalent rules
+	 */
+	public Set<Node> applyMapToGlobalURI( final Node node, final Map<Node, Set<TermMapping>> vocMap ) {
+		final Set<TermMapping> termMappings = vocMap.get(node);
+		final Set<Node> results = new HashSet<>();
+		if ( termMappings != null ) {
+			for (TermMapping mappingTypes : termMappings) {
+				if ( mappingTypes.getTypeOfRule().equals(OWL.equivalentClass.asNode()) || mappingTypes.getTypeOfRule().equals(OWL.equivalentProperty.asNode()) ) {
+					results.addAll(mappingTypes.getTranslatedTerms());
+				}
```

Comment:
Hi, I have fixed the issue regarding cases where no terms are added to the results (in [commit](https://github.com/LiUSemWeb/HeFQUIN/pull/298/commits/086062958db6bf390603ce9f2342ceaf9c694306)).

To avoid this issue, I added the original binding to the results if no mapping rules can be applied. 
In addition, I didn't include an else branch since the g2lMap contains all different types of term mappings. If a mapping rule does not fall within the equivalence types, it should be skipped rather than triggering an exception.

Reply:
> In addition, I didn't include an else branch since the g2lMap contains all different types of term mappings. If a mapping rule does not fall within the equivalence types, it should be skipped rather than triggering an exception.

I see your reasoning. However, the only cases in which this whole function (`applyMapToGlobalURI`)  should be called (according to the rest of your implementation) is from `applyMapToSolutionMapping` with `useInverse=false`. And that one should be called only from the G2L operator, and the G2L operator is safe only for equivalence-only vocab.mappings. Therefore, I was suggesting to throw the exception as an additional safety measure, just to make sure that we will see error messages if we use these functions in the future in a different way than currently anticipated.